File: cs50/PSET6/sentimental-credit/credit.py
# import get_string
from cs50 import get_string

# input should be num
while True:
    cardnum = get_string("Number: ")
    if cardnum.isnumeric():
        break

evens = 0
odds = 0

# reverse card numbers
card = reversed([int(digit) for digit in cardnum])

# Luhn's algorithm
for i, digit in enumerate(card):
    if (i + 1) % 2 == 0:
        digitodd = digit * 2
        if digitodd > 9:
            odds += int(digitodd / 10) + digitodd % 10
        else:
            odds += digitodd
    else:
        evens += digit

sumf = evens + odds
# variables
start = int(cardnum[0:2])
clen = len(cardnum)
sld = sumf % 10

# each card's algorithm
if sld == 0:
    if start in [34, 37] and clen == 15:
        print("AMEX")
    elif start in range(51, 56) and clen == 16:
        print("MASTERCARD")
    elif (int(cardnum[0]) == 4) and clen in [13, 16]:
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: cs50/PSET1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long n = get_long("Number: ");
    int i = 0;
    long count = n;
    while (count > 0)
    {
        count = count / 10;
        i++;
    }
    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    int sum1 = 0;
    int sum2 = 0;
    int mod1;
    int mod2;
    int d1;
    int d2;
    long x = n;
    int total = 0;
    do
    {
        mod1 = x % 10;
        x = x / 10;
        sum1 += mod1;

        mod2 = x % 10;
        x = x / 10;
        mod2 = mod2 * 2;
        d1 = mod2 % 10;
        d2 = mod2 / 10;
        sum2 = sum2 + d1 + d2;
    }
    while (x > 0);

    total = sum1 + sum2;
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    long f2 = n;
    do
    {
        f2 = f2 / 10;
    }
    while (f2 > 100);

    if (i == 15)
    {
        if (f2 == 34 || f2 == 37)
        {
            printf("AMEX\n");
        }
    }
    if (i == 16)
    {
        if (f2 == 51 || f2 == 52 || f2 == 53 || f2 == 54 || f2 == 55)
        {
            printf("MASTERCARD\n");
        }
    }

    long f1 = n;
    do
    {
        f1 = f1 / 10;
    }
    while (f1 > 10);

    if (i == 13 || i == 16)
    {
        if (f1 == 4)
        {
            printf("VISA\n");
        }
    }
}
